sram.c

```diff
@@ -0,0 +1,96 @@
+#include "sram.h"
+#include "spi.h"
+#include "nuts_bolts.h"
+
+/* Least and most significant bytes of x, of type uint16_t */
+#define LSB(x) (x & 0x00FF)
+#define MSB(x) ((x & 0xFF00) >> 8)
+
+#define MODE_MASK 0xC0
+#define MODE_IDX  6U
+
+/* SPI insructions */
+enum instruction_e {
+  READ = 0x3,
+  WRITE = 0x2,
+  EDIO = 0x3B,
+  EQIO = 0x38,
+  RSTIO = 0xFF,
+  RDMR = 0x5,
+  WRMR = 0x1
+};
+
+void sram_init()
+{
+  /* Configure SCS pin as output */
+  SCS_SRAM_DDR |= (1 << SCS_SRAM_DDR_PIN);
+
+  /* Set SCS pin to high. The CS for this chip is active low */
+  SCS_SRAM_PORT &= ~(1 << SCS_SRAM_PIN);
+
+  /* SCK resting state is 0. Clock data on rising edge */
+  spi_set_mode(0, 0);
+
+  sram_set_mode(BYTE_MODE);
+}
+
+uint8_t sram_read_mode()
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {RDMR, 0xFF};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+
+  return (data_in[1] & MODE_MASK) >> MODE_IDX;
+}
+
+void sram_set_mode(enum sram_mode_e mode)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {WRMR, mode << MODE_IDX};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+}
+
+uint8_t sram_read_byte(uint16_t addr)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[4] = {READ,
+                         MSB(addr),
+                         LSB(addr),
+                         0xFF};
```

Comment:
I agree with what you're saying, but I think it's better to declare the `data_out` array outside of the helper and send `data_out` to the helper function. The reason being that it's easier to compare this array with the data sheet's timing diagrams if you can see the four bytes that are transacted next to each other in the code.

I like your idea, but would tweak it as follows:

```
uint8_t _sram_helper(uint8_t * data_out)
{
...
}
```
```
uint8_t sram_read_byte(uint16_t addr)
{
  data_out[4] = {READ, MSB(addr), LSB(addr), 0xFF);
  return sram_helper(data_out);
}
```
etc.

(0xFF is a dummy byte that we need to send in order to receive the value that we're trying to read. SPI is inherently full-duplex, so you need to send a byte to receive a byte)